### PR TITLE
Add QCEP-7: PoSW diff adjustment

### PR DESCRIPTION
**Pull request by ninjaahhh**

### CODE REVIEW

**Comment by ninjaahhh** [comment]
bump up for reviews

File: QCEP/qcep-7.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+author: @qizhou, @junjiah
+status: Draft
+created: 2021-01-29
+---
+
+## Simple Summary and Motivation
+
+To better secure QuarkChain mainnet and enable bridging assets from other chains, we propose a self-adjusting mechanism to increase the network's hash power by lowering the effective difficulty for root chain PoSW stakers.
+
+The targeted network difficulty will match Ethereum mainnet's as of the time of writing, which is around 12.52 P (source: [2miners](https://2miners.com/eth-network-difficulty)). To compare, QuarkChain mainnet's root chain difficulty is ~91 T, roughly 1/100 of the former.
```
Comment (qizhou) [inline QCEP/qcep-7.md]: The current diff is
92,980,657,535,890
which means the hash power is 92,980,657,535,890 / 60s / 1e12 = 1.54 Th/s

File: QCEP/qcep-7.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+1.  `TARGETED_DIFF`: intended difficulty to match
+2. `ADJUSTING_STEP`: multiplier for `DIFF_DIVIDER`, the higher the number, the lower the effective difficulty will become as the time goes. For example, with initial difficulty 1 T, current `DIFF_DIVIDER` equal to 1000 and `ADJUSTING_STEP`equal to 10:
+   1. effective difficulty is 1 G, `DIFF_DIVIDER` increases 10 folds to 10k, thus network difficulty will stabilize at 10 T
+   2. time goes on (until target difficulty is met), `DIFF_DIVIDER` goes to 100k, and network difficulty stabilizes at 100 T
+   3. continue until `TARGETED_DIFF` is reached
+3. `ADJUSTING_INTERVAL`: number of blocks between adjustments
+
+## Rationale
+
+Those 3 parameters will lead to smoother difficulty adjustments.
```
Comment (qizhou) [inline QCEP/qcep-7.md]: We need an additional parameter of a start block?

Actually, I don't like TARGETED_DIFF parameter as if after the TARGETED_DIFF is reached, if the hash rate decreases (e.g., some machines are done), then DIFF_DIVIDER can further increase but never decreases after the hash rate recovers, which is out of our expectation.

Firstly, I think we should have a consistent prefix of parameters for the upgrade, e.g., BOOST

- BOOST_TIMESTAMP:  timestamp to enable boost (first-time increase DIFF_DIVIDER)
- BOOST_MULTIPLIER_PER_STEP: 10x  
- BOOST_STEPS:  5 (i.e., 10x ** 5, target is 1.54 Th/s * 10000 => 15.4 Ph/s)
- BOOST_STEPS_INTERVAL: interval of applying next MULTIPLER

Therefore the new DIFF_DIVIDER equation becomes
```
if block.timestamp >= BOOST_TIMESTAMP:
    DIFF_DIVIDER *= BOOST_MULTIPLIER_PER_STEP ** min(((block.timestamp - BOOST_TIMESTAMP) // BOOST_STEP_INTERVAL) + 1, BOOST_STEPS)
```
which can be computed very efficiently.

File: QCEP/qcep-7.md [hunk 3]
```diff
@@ -1,31 +1,35 @@
...
-In the configuration for full nodes, will have three more fields in `POSW_CONFIG`:
+In the configuration for full nodes, will have four more fields in `POSW_CONFIG`:
 
-1.  `TARGETED_DIFF`: intended difficulty to match
-2. `ADJUSTING_STEP`: multiplier for `DIFF_DIVIDER`, the higher the number, the lower the effective difficulty will become as the time goes. For example, with initial difficulty 1 T, current `DIFF_DIVIDER` equal to 1000 and `ADJUSTING_STEP`equal to 10:
-   1. effective difficulty is 1 G, `DIFF_DIVIDER` increases 10 folds to 10k, thus network difficulty will stabilize at 10 T
-   2. time goes on (until target difficulty is met), `DIFF_DIVIDER` goes to 100k, and network difficulty stabilizes at 100 T
-   3. continue until `TARGETED_DIFF` is reached
-3. `ADJUSTING_INTERVAL`: number of blocks between adjustments
+1. `BOOST_TIMESTAMP`: timestamp to enable boost (first-time increase DIFF_DIVIDER)
```
Comment (qizhou) [inline QCEP/qcep-7.md]: Should we explain the disabled case, i.e., BOOST_TIMESTAMP = 0?

File: QCEP/qcep-7.md [hunk 4]
```diff
@@ -1,31 +1,35 @@
...
+2. `BOOST_MULTIPLIER`_PER_STEP: multiplier for `DIFF_DIVIDER` for each step (i.e. 2x)
```
Comment (qizhou) [inline QCEP/qcep-7.md]: `BOOST_MULTIPLIER`_PER_STEP => `BOOST_MULTIPLIER_PER_STEP`

**Review by qizhou — COMMENTED** [review]
We should also list the mainnet configuration in the qcep?

File: QCEP/qcep-7.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+In the configuration for full nodes, will have four more fields in `POSW_CONFIG`:
+
+1. `BOOST_TIMESTAMP`: timestamp to enable boost (first-time increase DIFF_DIVIDER), set BOOST_TIMESTAMP to 0 mean disable this feature;
+2. `BOOST_MULTIPLIER_PER_STEP`: multiplier for `DIFF_DIVIDER` for each step (i.e. 2x);
+3. `BOOST_STEPS`: number of steps to increase DIFF_DIVIDER (i.e. 10);
+4. `BOOST_STEP_INTERVAL`: interval of applying next MULTIPLIER.
+
+Therefore the new DIFF_DIVIDER equation becomes
+```
+if block.timestamp >= BOOST_TIMESTAMP:
```
Comment (qizhou) [inline QCEP/qcep-7.md]: The code should be updated accordingly?

**Comment by qizhou** [comment]
> We should also list the mainnet configuration in the qcep?

I mean the actual parameters should be listed here?

**Review by qizhou — APPROVED** [review]

**Pull request merged**